### PR TITLE
NCP Usage of otLinkRaw API

### PR DESCRIPTION
**Pull request by nibanks**
This updates the NCP code to expose the otLinkRaw interface via Spinel commands. The new code can be compile time removed, but cannot (yet) be configured to compile down to _only_ the raw link-layer interface. Adds a new Spinel capability to reflect support for the raw link-layer interface.

Note, for some reason the Linux builds are failing, but I am not sure why. Any help with those would be appreciated.

### CODE REVIEW

**Comment by nibanks** [comment]
The Linux builds are failing as if they aren't including the declarations correctly. I have tried a number of things like moving the include header around but it never seems to find the definitions of the functions. See the failures [here](https://travis-ci.org/nibanks/openthread/builds/194944940).

**Comment by jwhui** [comment]
It appears to be a linker issue.  I would try swapping the order of the [NCP LDADD list](https://github.com/openthread/openthread/blob/master/examples/apps/ncp/Makefile.am#L41).

**Comment by nibanks** [comment]
Thanks @jwhui! That seemed to fix my Travis builds.

**Comment by nibanks** [comment]
I have additional changes to add support for calling the energy scan raw link layer interface. Would you prefer that in a separate PR after this one, or just add it on to this one?

**Comment by jwhui** [comment]
Generally, it's easier and quicker to review smaller changes.  That said, if the update to this PR is relatively small, you could apply those changes to this PR.

**Comment by nibanks** [comment]
Ok, I think it would be best if we can get a quick review of this code and then I will send a new PR for the energy scan stuff.

**Comment by nibanks** [comment]
Any comments on this PR or is everyone fine as is? I still have two more PR queued up waiting on this one.

File: src/ncp/ncp_base.cpp [hunk 1]
```diff
@@ -239,6 +241,9 @@ const NcpBase::SetPropertyHandlerEntry NcpBase::mSetPropertyHandlerTable[] =
     { SPINEL_PROP_MAC_SCAN_PERIOD, &NcpBase::SetPropertyHandler_MAC_SCAN_PERIOD },
     { SPINEL_PROP_MAC_15_4_PANID, &NcpBase::SetPropertyHandler_MAC_15_4_PANID },
     { SPINEL_PROP_MAC_RAW_STREAM_ENABLED, &NcpBase::SetPropertyHandler_MAC_RAW_STREAM_ENABLED },
+#if OPENTHREAD_ENABLE_RAW_LINK_API
+    { SPINEL_PROP_STREAM_RAW, &NcpBase::SetPropertyHandler_STREAM_RAW },
+#endif // OPENTHREAD_ENABLE_RAW_LINK_API
```
Comment (turon) [inline src/ncp/ncp_base.cpp]: Is it possible to group RAW_LINK ifdefs together, or does order matter here?

File: src/ncp/ncp_base.cpp [hunk 2]
```diff
@@ -481,6 +486,10 @@ NcpBase::NcpBase(otInstance *aInstance):
     mRequireJoinExistingNetwork(false),
     mIsRawStreamEnabled(false),
     mDisableStreamWrite(false),
+#if OPENTHREAD_ENABLE_RAW_LINK_API
+    mCurTransmintTID(0),
```
Comment (turon) [inline src/ncp/ncp_base.cpp]: typo / style nit: mCurTransmitTid (or mCurrentTransmitTid or mCurrentTxTid)?

Though minty TIDs sound tasty ;)

File: src/ncp/ncp_base.cpp [hunk 3]
```diff
@@ -748,6 +757,99 @@ void NcpBase::HandleEnergyScanResult(otEnergyScanResult *aResult)
...
+}
+
+void NcpBase::LinkRawReceiveDone(RadioPacket *aPacket, ThreadError aError)
+{
+    ThreadError errorCode = kThreadError_None;
+    uint16_t flags = 0;
+
+    SuccessOrExit(errorCode = OutboundFrameBegin());
+
+    if (aPacket->mDidTX)
```
Comment (turon) [inline src/ncp/ncp_base.cpp]: style nit: mDidTx

File: src/ncp/ncp_base.cpp [hunk 4]
```diff
@@ -748,6 +757,99 @@ void NcpBase::HandleEnergyScanResult(otEnergyScanResult *aResult)
...
+            )
+        );
+    }
+
+    // Append metadata (rssi, etc)
+    SuccessOrExit(
+        errorCode = OutboundFrameFeedPacked(
+            "ccSiCC",
+            aPacket->mPower,    // TX Power
+            -128,               // Noise Floor (Currently unused)
```
Comment (turon) [inline src/ncp/ncp_base.cpp]: Magic

File: src/ncp/ncp_base.cpp [hunk 5]
```diff
@@ -3066,11 +3181,23 @@ ThreadError NcpBase::SetPropertyHandler_PHY_ENABLED(uint8_t header, spinel_prop_
     {
         if (value == false)
         {
-            errorCode = otPlatRadioDisable(mInstance);
+            // If we have raw stream enabled stop receiving
+            if (mIsRawStreamEnabled)
+            {
+                otLinkRawSleep(mInstance);
```
Comment (turon) [inline src/ncp/ncp_base.cpp]: Previous semantics just turned the radio off, is it correct to sleep here?

File: src/ncp/ncp_base.cpp [hunk 6]
```diff
@@ -3119,6 +3248,25 @@ ThreadError NcpBase::SetPropertyHandler_PHY_CHAN(uint8_t header, spinel_prop_key
...
         errorCode = otSetChannel(mInstance, static_cast<uint8_t>(i));
 
+#if OPENTHREAD_ENABLE_RAW_LINK_API
+
+         if (errorCode == kThreadError_None)
+         {
+             // Cache the channel. If the raw link layer isn't enabled yet, the otSetChannel call
+             // doesn't call into the radio layer to set the channel. We will have to do it
+             // manually whenever the radios are enabled and/or raw stream is enabled.
+             mCurReceiveChannel = static_cast<uint8_t>(i);
```
Comment (turon) [inline src/ncp/ncp_base.cpp]: Suggest renaming local variable `i` to `channel` or `ch` for readability.

File: src/ncp/ncp_base.cpp (hunk 5 above)
Comment (jwhui) [inline src/ncp/ncp_base.cpp]: As discussed in #1170, default state is Sleep.

File: src/ncp/ncp_base.cpp (hunk 6 above)
Comment (jwhui) [inline src/ncp/ncp_base.cpp]: Using `i` seems consistent with the rest of the NCP code - I wouldn't worry about it in this PR.

File: src/ncp/ncp_base.cpp (hunk 4 above)
Comment (jwhui) [inline src/ncp/ncp_base.cpp]: This is copied from elsewhere in NCP and (as noted) is currently unused.  I wouldn't worry about it here.

File: src/ncp/ncp_base.cpp [hunk 7]
```diff
@@ -748,6 +757,99 @@ void NcpBase::HandleEnergyScanResult(otEnergyScanResult *aResult)
...
+}
+
+void NcpBase::LinkRawTransmitDone(otInstance *, RadioPacket *aPacket, bool aFramePending, ThreadError aError)
+{
+    sNcpContext->LinkRawTransmitDone(aPacket, aFramePending, aError);
+}
+
+void NcpBase::LinkRawTransmitDone(RadioPacket *, bool aFramePending, ThreadError aError)
+{
+    if (mCurTransmintTID)
```
Comment (abtink) [inline src/ncp/ncp_base.cpp]: nit: `mCurTransmintTID` -> `mCurTransmitTID`

File: src/ncp/ncp_base.cpp [hunk 8]
```diff
@@ -3418,6 +3579,69 @@ ThreadError NcpBase::SetPropertyHandler_MAC_RAW_STREAM_ENABLED(uint8_t header, s
     return errorCode;
 }
 
+#if OPENTHREAD_ENABLE_RAW_LINK_API
+
+ThreadError NcpBase::SetPropertyHandler_STREAM_RAW(uint8_t header, spinel_prop_key_t key, const uint8_t *value_ptr,uint16_t value_len)
```
Comment (abtink) [inline src/ncp/ncp_base.cpp]: nit: space after `,` in `value_ptr,uint..`

File: src/ncp/ncp_base.cpp (hunk 1 above)
Comment (nibanks) [inline src/ncp/ncp_base.cpp]: Practically I don't think order matters. If you guys want, I can group these.

File: src/ncp/ncp_base.cpp (hunk 2 above)
Comment (nibanks) [inline src/ncp/ncp_base.cpp]: I can change to `mCurrentTransmitTid`.

File: src/ncp/ncp_base.cpp (hunk 3 above)
Comment (nibanks) [inline src/ncp/ncp_base.cpp]: This was an already existing parameter in the `RadioPacket` class.

Comment (jwhui) [inline src/ncp/ncp_base.cpp]: @nibanks, agree - don't worry about it in this PR.

File: src/ncp/ncp_base.cpp (hunk 8 above)
Comment (nibanks) [inline src/ncp/ncp_base.cpp]: I will fix this.

File: src/ncp/ncp_base.cpp (hunk 1 above)
Comment (jwhui) [inline src/ncp/ncp_base.cpp]: @abtink, thoughts?

Comment (abtink) [inline src/ncp/ncp_base.cpp]: I am ok with both ways.
I like combining them maybe a bit more (to keep the code cleaner)...

File: src/ncp/ncp_base.cpp (hunk 3 above)
Comment (abtink) [inline src/ncp/ncp_base.cpp]: Yes, let's keep it...

**Comment by jwhui** [comment]
It appears Appveyor did not pick up your last push...

**Comment by nibanks** [comment]
I noticed that, though I don't know why. My [branch's build](https://ci.appveyor.com/project/nibanks/openthread/build/0.1.597) built fine though. I'd be ok merging anyways.

**Pull request merged**